### ISSUE TITLE
Exception occurs when converting object to Decimal using ClassData

### ISSUE THREAD

**Issue by bmeredith** [xunit/xunit#1771]
# Expected Behavior

The below test should pass

Note: Test passes in versions of xunit < 2.4.0

# Actual Behavior

Test fails to pass because of an exception being thrown from xUnit.

Stacktrace:
```
System.Reflection.AmbiguousMatchException: Ambiguous match found.
   at System.DefaultBinder.SelectMethod(BindingFlags bindingAttr, MethodBase[] match, Type[] types, ParameterModifier[] modifiers)
   at System.RuntimeType.GetMethodImplCommon(String name, Int32 genericParameterCount, BindingFlags bindingAttr, Binder binder, CallingConventions callConv, Type[] types, ParameterModifier[] modifiers)
   at System.RuntimeType.GetMethodImpl(String name, BindingFlags bindingAttr, Binder binder, CallingConventions callConv, Type[] types, ParameterModifier[] modifiers)
   at System.Type.GetMethod(String name, Type[] types)
   at System.Reflection.RuntimeReflectionExtensions.GetRuntimeMethod(Type type, String name, Type[] parameters)
   at Xunit.Sdk.TypeUtility.PerformDefinedConversions(Object argumentValue, Type conversionDeclaringType) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\TypeUtility.cs:line 169
   at Xunit.Sdk.TypeUtility.TryConvertObject(Object argumentValue, Type parameterType) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\TypeUtility.cs:line 145
   at Xunit.Sdk.TypeUtility.ResolveMethodArguments(MethodBase testMethod, Object[] arguments) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\TypeUtility.cs:line 115
   at Xunit.Sdk.TestMethodTestCase.Initialize() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestMethodTestCase.cs:line 289
--- End of stack trace from previous location where exception was thrown ---
   at Xunit.Sdk.TestInvoker`1.<RunAsync>b__47_0() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 206
   at Xunit.Sdk.ExceptionAggregator.RunAsync[T](Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 107
```

# Steps to Reproduce the Problem

1. Create a new xUnit project
2. Paste the following code in the created project:
```
    public class Class
    {
        public object Value { get; set; }
    }

    public class Test
    {
        [Theory, ClassData(typeof(Data))]
        public void GetValue<T>(T value)
        {
            new Class
            {
                Value = value
            };
        }

        private class Data : IEnumerable<object[]>
        {
            private readonly List<object[]> _data = new List<object[]>
            {
                new object[] { 1.0m }
            };

            public IEnumerator<object[]> GetEnumerator()
            {
                return _data.GetEnumerator();
            }

            IEnumerator IEnumerable.GetEnumerator()
            {
                return GetEnumerator();
            }
        }
    }
```
3. Run the test

# Versions

- xunit: 2.4.0
- xunit.runner.visualstudio: 2.4.0

**Comment by joaonunomota** [comment]
This issue is fixed in version 2.4.2, here are the before and after results of each version of the package:

Before: (2.4.1)

![image](https://github.com/xunit/xunit/assets/16689477/006b1291-14cc-4de0-8fd2-d4ff13cae0c2)

After: (2.4.2)

![image](https://github.com/xunit/xunit/assets/16689477/49a10b4f-708d-4996-9ab9-93eac0de3384)

The test class used is a combination of the example in this issue and in issue #2298.

```
using System.Collections;
using Xunit;

namespace example;

public class ClassDataTests
{
    [Theory, ClassData(typeof(IntegerLiteralExample))]
    public void IntegerLiteralTest<T>(T value)
    {
        new Class
        {
            Value = value
        };
    }

    [Theory, ClassData(typeof(GuardClausesExample))]
    public void GuardClausesTest<T>(T input, Func<T, bool> func)
    {
        Assert.True(func(input));
    }

    private class IntegerLiteralExample : IEnumerable<object[]>
    {
        private readonly List<object[]> _data = new List<object[]>
            {
                new object[] { 1.0m },
                new object[] { (uint)2 },
                new object[] { (nuint)3 },
            };

        public IEnumerator<object[]> GetEnumerator()
        {
            return _data.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }

    private class GuardClausesExample : IEnumerable<object[]>
    {
        public IEnumerator<object[]> GetEnumerator()
        {
            yield return new object[] { 20.0m, (Func<decimal, bool>)((x) => x > 10.0m) };
        }
        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
    public class Class
    {
        public object Value { get; set; }
    }
}
```

**Comment by bradwilson** [comment]
Thanks @joaonunomota!
